Add RPC command shield_coinbase #2448.

bitcartel opened the pull request:
Shield transparent coinbase funds by sending to a shielded z address. Closes #2448.

bitcartel commented:
Tests added.  Ready for review.

str4d commented on src/wallet/asyncrpcoperation_shieldcoinbase.cpp:
```diff
@@ -0,0 +1,451 @@
+        UniValue contextInfo) :
+        inputs_(inputs), fee_(fee), contextinfo_(contextInfo)
+{
+    if (fee_ < 0 || fee > MAX_MONEY) {
```
Why does this check two different parameters (one derived from the other, meaning there could be some invisible-to-here divergence between the two)?

bitcartel replied:
Fixed typo, should have been fee.

str4d commented on src/wallet/asyncrpcoperation_shieldcoinbase.cpp:
```diff
@@ -0,0 +1,451 @@
+    uint256 anchor;
+    {
+        LOCK2(cs_main, pwalletMain->cs_wallet);
+        anchor = pcoinsTip->GetBestAnchor();    // As there are no inputs, ask the wallet for the best anchor
```
Wallet shouldn't be involved in this, just the coins view. So do we need to lock on it?

bitcartel replied:
Removed. It was inherited from z_sendmany and I filed a ticket about it.

str4d commented on src/wallet/asyncrpcoperation_shieldcoinbase.cpp:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    if (!(witnesses.size() == info.notes.size())) {
+        throw runtime_error("number of notes and witnesses do not match");
```
Not as clear as checking that `info.notes.size() == 0`, given that in this method's context there will never be input notes. Or are you planning for some later refactoring?

bitcartel replied:
Legacy from z_sendmany.  Removed both witnesses and notes since not relevant to shielding from pure taddr.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3494,6 +3495,178 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+            continue;
+        }
+
+        bool isCoinbase = out.tx->IsCoinBase();
```
Non-blocking: Variable is only used once, could be inlined.

bitcartel replied:
Done

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3494,6 +3495,178 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+        if (!maxedOutFlag) {
+            CBitcoinAddress ba(address);
+            size_t increase = (ba.IsScript()) ? CTXIN_SPEND_P2SH_SIZE : CTXIN_SPEND_DUST_SIZE;
+            if (estimatedTxSize + increase >= MAX_TX_SIZE || utxoCounter == mempoolLimit) {
```
This defines "maxing out" as having exceeded, rather than reaching, `MAX_TX_SIZE`, will possibly cause transactions to be attempted that exceed the maximum size, depending on the particular makeup. In practice it should be fine, as the calculations use conservative estimates with wiggle-room.

bitcartel replied:
Good spot.  Fixed definition of "maxing out" to be when MAX_TX_SIZE is reached.  There is actually another subtle bug where utxoCounter should only be checked when mempoolLimit > 0 ; this has also been fixed.

bitcartel commented:
@str4d I addressed review comments. Force pushed to keep a clean commit.  Here are the changes based on your review: https://gist.github.com/bitcartel/28eecca6220cb18f8a84a83cb5f7c680

bitcartel commented on qa/rpc-tests/wallet_shieldcoinbase.py:
```diff
@@ -0,0 +1,185 @@
+        assert_equal(self.nodes[1].getbalance(), 20)
+        assert_equal(self.nodes[2].getbalance(), 30)
+
+        # Shield coinbase utxos from any node 2 taddr, and set fee to 0
```
Done

bitcartel commented on src/test/rpc_wallet_tests.cpp:
```diff
@@ -1214,29 +1216,169 @@ BOOST_AUTO_TEST_CASE(rpc_wallet_encrypted_wallet_zkeys)
     BOOST_CHECK_NO_THROW(retValue = CallRPC("z_listaddresses"));
     arr = retValue.get_array();
-    BOOST_CHECK(arr.size() == n+1);    
+    BOOST_CHECK(arr.size() == n+1);
```
:smiley:

bitcartel commented:
Here's an example using `z_shieldcoinbase` on testnet to shield coinbase utxos (multi-sig):
https://explorer.testnet.z.cash/block/000e638de52dae3dd61066d5eabff301ea86bbca42aced6647878062df1152c1

arcalinea commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3494,6 +3495,182 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+    if (fHelp || params.size() < 2 || params.size() > 3)
+        throw runtime_error(
+            "z_shieldcoinbase \"fromaddress\" \"tozaddress\" ( fee )\n"
+            "\nShield transparent coinbase funds by sending to a shielded z address.  This is an asynchronous operation and utxos"
```
Other z prefix rpc calls use the term "zaddr" in the cli output, and it is used this way farther down. Should keep this consistent, so change "z address" to "zaddr"

bitcartel replied:
Done.

arcalinea commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3494,6 +3495,182 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+
+    // Validate the from address
+    auto fromaddress = params[0].get_str();
+    bool isFromWildcard = fromaddress == "*";
```
Is there an advantage to having users explicitly pass a wildcard if they don't care which taddr they're shielding from, as opposed to just leaving it as an optional secondary parameter than can be omitted? It seems more intuitive to me to leave it as an optional parameter, like the fee, that defaults to a wildcard but can be set to a specific address.

bitcartel replied:
It fits in with upstream / inherited RPC calls where the user must explicitly pass in a wildcard.  In future, when have full RPC 2.0 support and we pass in key-value pairs, we'll have more flexibility with regards to optional parameters and defaults.

arcalinea replied:
Ah ok, sounds good.

arcalinea commented:
ACK, left some non-blocking questions and comments. Tests pass for me. Will also try to pull this and test on the testnet faucet this afternoon and see how it performs, this is a sorely needed feature. :)

daira commented on src/wallet/asyncrpcoperation_shieldcoinbase.cpp:
```diff
@@ -0,0 +1,441 @@
+  #else
+    GenerateBitcoins(false, 0);
+  #endif
+#endif
```
This (and the reenablement) seems like it should be factored out if it is intended to apply to all async operations.

daira commented on src/wallet/asyncrpcoperation_shieldcoinbase.cpp:
```diff
@@ -0,0 +1,441 @@
+}
+
+
+UniValue AsyncRPCOperation_shieldcoinbase::perform_joinsplit(ShieldCoinbaseJSInfo & info) {
```
This code seems like it should be shared with ``z_sendmany``.

daira commented on src/wallet/asyncrpcoperation_shieldcoinbase.h:
```diff
@@ -0,0 +1,122 @@
+    std::vector<JSOutput> vjsout;
+    CAmount vpub_old = 0;
+    CAmount vpub_new = 0;
+};
```
This also seems like it should be shared with ``z_sendmany``.

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3494,6 +3495,182 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+            FormatMoney(shieldedValue), FormatMoney(nFee)));
+    }
+
+    // Check that the user specified fee is sane (if too high, it can result in error -25 absurd fee)
```
"is acceptable"
